mqtt-v5/src/decoder.rs

```diff
@@ -405,7 +408,8 @@ fn decode_connect(bytes: &mut Cursor<&mut BytesMut>) -> Result<Option<Packet>, D
             })?);
         }
 
-        let topic = read_string!(bytes);
+        let topic = Topic::from_str(read_string!(bytes).as_str())
+            .map_err(|_| DecodeError::InvalidTopicFilter(TopicParseError::TopicTooLong))?;
```

Comment:
`Topic::from_str()` returns a `Result` with `TopicParseError` being the error type. This function returns a `DecodeError` so you're right to map the error and create a `DecodeError`. However, we already have the `TopicParseError` from the call to `from_str`, so we can use that directly:

```suggestion
        let topic =
            Topic::from_str(read_string!(bytes).as_str()).map_err(DecodeError::InvalidTopic)?;
```

The other thing to note: this is the topic for the final will, it's something that _publishes_ but doesn't _subscribe_. Topic filters are only for subscribers, so we want to use `InvalidTopic` instead of `InvalidTopicFilter`

Reply:
Thanks, that Error Message makes much more sense